presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -460,6 +472,50 @@ private void checkTypesMatchForInsert(Insert insert, Scope queryScope, List<Colu
             }
         }
 
+        private void checkTypesMatchForNestedStructs(
+                Node node,
+                String errorMessage,
+                int i,
+                List<Type> columnRowTypes,
+                List<RowType.Field> expectedRowFields,
+                String fieldName)
+        {
+            if (columnRowTypes.size() != expectedRowFields.size()) {
+                throw new SemanticException(MISMATCHED_SET_COLUMN_TYPES,
+                        node,
+                        errorMessage + "Mismatch at column %d: '%s' has %d field(s) but expression has %d field(s)",
+                        i + 1,
+                        fieldName,
+                        expectedRowFields.size(),
+                        columnRowTypes.size());
+            }
+            for (int j = 0; j < columnRowTypes.size(); j++) {
+                if (!metadata.getFunctionAndTypeManager().canCoerce(
+                        columnRowTypes.get(j),
+                        expectedRowFields.get(j).getType())) {
+                    fieldName += "." + expectedRowFields.get(j).getName().get();
+                    if (columnRowTypes.get(j) instanceof RowType && expectedRowFields.get(j).getType() instanceof RowType) {
+                        checkTypesMatchForNestedStructs(
+                                node,
+                                errorMessage,
+                                i,
+                                columnRowTypes.get(j).getTypeParameters(),
+                                ((RowType) expectedRowFields.get(i).getType()).getFields(),
```

Comment:
It might be a dumb question, but is using `j` for `columnRowTypes` but `i` for `expectedRowFields` correct?

Reply:
You're right. This was an error